django-s3-file-field-client/src/client.ts

```diff
@@ -25,10 +25,19 @@ export interface UploadResult {
   state: 'aborted' | 'successful' | 'error';
 }
 
-export default class S3FFClient {
-  protected readonly baseUrl: string;
+export interface ProgressEvent {
+  loaded?: number;
+  total?: number;
+  state: 'initializing' | 'sending' | 'finalizing';
+}
+
+type ProgressCallback = (progress: ProgressEvent) => void;
 
-  constructor(baseUrl: string) {
+export default class S3FFClient {
+  constructor(
+    protected readonly baseUrl: string,
+    private readonly onProgress: ProgressCallback = () => { /* no-op*/ },
```

Comment:
Do you think it'd be helpful to ever know when there's no callback attached?

If so, we should make `onProgress` default to `undefined`.

I'm not sure if this is something we'd ever care about though. It's just a question, not a suggestion.

Reply:
I think this default value is an internal detail, so if it ever becomes a requirement for us to know whether it was passed in or not, we can change it without any issue. Absent that, using a non-optional function type greatly simplifies the implementation, so I'd prefer to leave it.

Reply:
Agreed.